lib/datadog/tracing/trace_segment.rb

```diff
@@ -49,27 +63,35 @@ def initialize(
         runtime_id: nil,
         sample_rate: nil,
         sampling_priority: nil,
-        service: nil
+        service: nil,
+        tags: nil,
+        metrics: nil
       )
         @id = id
         @root_span_id = root_span_id
         @spans = spans || []
-        @tags = {}
-
-        # Set well-known tags
-        self.agent_sample_rate = agent_sample_rate
-        self.hostname = hostname
-        self.lang = lang
-        self.name = (name.frozen? ? name : name.dup)
-        self.origin = (origin.frozen? ? origin : origin.dup)
-        self.process_id = process_id
-        self.rate_limiter_rate = rate_limiter_rate
-        self.resource = (resource.frozen? ? resource : resource.dup)
-        self.rule_sample_rate = rule_sample_rate
-        self.runtime_id = runtime_id
-        self.sample_rate = sample_rate
-        self.sampling_priority = sampling_priority
-        self.service = (service.frozen? ? service : service.dup)
+
+        # Does not make an effort to move metrics out of tags
+        # The caller is expected to have done that
+        @meta = tags || {}
+        @metrics = metrics || {}
+
+        # Set well-known tags, defaulting to getting the values from tags
+        @agent_sample_rate = agent_sample_rate || agent_sample_rate_tag
+        @hostname = hostname || hostname_tag
+        @lang = lang || lang_tag
+        @name = (name || name_tag).tap { |v| break (v.frozen? ? v : v.dup) }
```

Comment:
What does `break` do here?

Also there's a small performance cost paid for using `tap`. Would it be easier (even if more verbose) to do the following?

```
@name = name || name_tag
@name = (@name.frozen? ? @name : @name.dup)
```

It's more lines but easier to read and possibly slightly more performant.

Reply:
`tap` returns `self` normally, `break` breaks out of the block, making `tap` return `break`'s argument.

Reply:
Addressed by factoring frozen || dup to a Core::Util